### ISSUE TITLE
c++ demo

### ISSUE THREAD

**Issue by huaze555** [zju3dv/snake#84]
can anybody release a c++ demo？

**Comment by wuzuowuyou** [comment]
libtorch?

**Comment by huaze555** [comment]
@wuzuowuyou libtorch cannot export dcn_v2 ...

**Comment by wuzuowuyou** [comment]
@huaze555 

how can libtorch export dcn_v2 ?

**Comment by DesmondSL** [comment]
i change code in lib/networks/dcn_v2.py 

class _DCNv2(Function):
    @staticmethod
    def forward(ctx, input, offset, mask, weight, bias,
                stride, padding, dilation, deformable_groups):
        ctx.stride = _pair(stride)
        ctx.padding = _pair(padding)
        ctx.dilation = _pair(dilation)
        ctx.kernel_size = _pair(weight.shape[2:4])
        ctx.deformable_groups = deformable_groups
        output = _backend.dcn_v2_forward(input, weight, bias,
                                         offset, mask,
                                         ctx.kernel_size[0], ctx.kernel_size[1],
                                         ctx.stride[0], ctx.stride[1],
                                         ctx.padding[0], ctx.padding[1],
                                         ctx.dilation[0], ctx.dilation[1],
                                         ctx.deformable_groups)
        ctx.save_for_backward(input, offset, mask, weight, bias)
        return output

    @staticmethod
    @once_differentiable
    def backward(ctx, grad_output):
        input, offset, mask, weight, bias = ctx.saved_tensors
        grad_input, grad_offset, grad_mask, grad_weight, grad_bias = \
            _backend.dcn_v2_backward(input, weight,
                                     bias,
                                     offset, mask,
                                     grad_output,
                                     ctx.kernel_size[0], ctx.kernel_size[1],
                                     ctx.stride[0], ctx.stride[1],
                                     ctx.padding[0], ctx.padding[1],
                                     ctx.dilation[0], ctx.dilation[1],
                                     ctx.deformable_groups)

        return grad_input, grad_offset, grad_mask, grad_weight, grad_bias,\
            None, None, None, None,


dcn_v2_conv = _DCNv2.apply


class DCNv2(nn.Module):

    def __init__(self, in_channels, out_channels,
                 kernel_size, stride, padding, dilation=1, deformable_groups=1):
        super(DCNv2, self).__init__()
        self.in_channels = in_channels
        self.out_channels = out_channels
        self.kernel_size = _pair(kernel_size)
        self.stride = _pair(stride)
        self.padding = _pair(padding)
        self.dilation = _pair(dilation)
        self.deformable_groups = deformable_groups

        self.weight = nn.Parameter(torch.Tensor(
            out_channels, in_channels, *self.kernel_size))
        self.bias = nn.Parameter(torch.Tensor(out_channels))
        self.reset_parameters()

    def reset_parameters(self):
        n = self.in_channels
        for k in self.kernel_size:
            n *= k
        stdv = 1. / math.sqrt(n)
        self.weight.data.uniform_(-stdv, stdv)
        self.bias.data.zero_()

    # def forward(self, input, offset, mask):
    #     assert 2 * self.deformable_groups * self.kernel_size[0] * self.kernel_size[1] == \
    #         offset.shape[1]
    #     assert self.deformable_groups * self.kernel_size[0] * self.kernel_size[1] == \
    #         mask.shape[1]
    #     return dcn_v2_conv(input, offset, mask,
    #                        self.weight,
    #                        self.bias,
    #                        self.stride,
    #                        self.padding,
    #                        self.dilation,
    #                        self.deformable_groups)

    def forward(self, input, offset, mask):
        output = _backend.dcn_v2_forward(input, self.weight, self.bias,
                            offset, mask,
                            self.weight.shape[2], self.weight.shape[3],
                            self.stride[0], self.stride[1],
                            self.padding[0], self.padding[1],
                            self.dilation[0], self.dilation[1],
                            self.deformable_groups)
        return output


class DCN(DCNv2):

    def __init__(self, in_channels, out_channels,
                 kernel_size, stride, padding,
                 dilation=1, deformable_groups=1):
        super(DCN, self).__init__(in_channels, out_channels,
                                  kernel_size, stride, padding, dilation, deformable_groups)

        channels_ = self.deformable_groups * 3 * self.kernel_size[0] * self.kernel_size[1]
        self.conv_offset_mask = nn.Conv2d(self.in_channels,
                                          channels_,
                                          kernel_size=self.kernel_size,
                                          stride=self.stride,
                                          padding=self.padding,
                                          bias=True)
        self.init_offset()

    def init_offset(self):
        self.conv_offset_mask.weight.data.zero_()
        self.conv_offset_mask.bias.data.zero_()

    def forward(self, input):
        out = self.conv_offset_mask(input)
        o1, o2, mask = torch.chunk(out, 3, dim=1)
        offset = torch.cat((o1, o2), dim=1)
        mask = torch.sigmoid(mask)
        output = _backend.dcn_v2_forward(input, self.weight, self.bias,
                            offset, mask,
                            self.weight.shape[2], self.weight.shape[3],
                            self.stride[0], self.stride[1],
                            self.padding[0], self.padding[1],
                            self.dilation[0], self.dilation[1],
                            self.deformable_groups)
        return output
        # return dcn_v2_conv(input, offset, mask,
        #                    self.weight, self.bias,
        #                    self.stride,
        #                    self.padding,
        #                    self.dilation,
        #                    self.deformable_groups)

but i got trace warning as followed:

TracerWarning: Output nr 1. of the traced function does not match the corresponding output of the Python function. Detailed error:
Not within tolerance rtol=1e-05 atol=1e-05 at input[0, 0, 109, 64] (125.1081771850586 vs. 8.114267349243164) and 32766 other locations (99.00%)
  _check_trace([example_inputs], func, executor_options, traced, check_tolerance, _force_outplace)

**Comment by huaze555** [comment]
> @huaze555
> 
> how can libtorch export dcn_v2 ?

I don't know so i am here ask this question.
